### PR TITLE
Make statusPingTimeout configurable

### PR DESCRIPTION
**Pull request by yangcao77**
Signed-off-by: Stephanie <stephanie.cao@ibm.com>

This PR is for issue https://github.com/eclipse/codewind/issues/1299

make `statusPingTimeout` configurable

Also added unit test for the new project specification setting:
```
      combinational testing of changeStatusPingTimeout function
        ✓ combo1 => statusPingTimeoutSettings: {"statusPingTimeout":"10"}
        ✓ combo2 => statusPingTimeoutSettings: {"statusPingTimeout":"test"}
```

### CODE REVIEW

**Comment by yangcao77** [comment]
@rajivnathan Can you help review this PR?

File: src/pfe/file-watcher/server/src/controllers/projectStatusController.ts [hunk 1]
```diff
@@ -56,13 +57,14 @@ const pingCountMap = new Map();
...
  */
-class ProjectState {
+export class ProjectState {
     state: any;
     msg: string;
     lastbuild?: number;
     appImageLastBuild?: string;
     buildImageLastBuild?: string;
     detailedAppStatus?: DetailedAppStatus;
+    isFirstTimePing?: boolean;
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply src/pfe/file-watcher/server/src/controllers/projectStatusController.ts]: We still update `detailedAppStatus` if the value is false,  we just not sending any pinging messages.

File: src/pfe/file-watcher/server/src/projects/projectSpecifications.ts [hunk 2]
```diff
@@ -797,6 +801,61 @@ const changeMavenProperties = async function (mavenProperties: any, operation: O
...
+ */
+const changeStatusPingTimeout = async function (statusPingTimeout: string, operation: Operation): Promise<any> {
+
+    const projectInfo: ProjectInfo = operation.projectInfo;
+    const projectID = projectInfo.projectID;
+    const projectHandler = await projectExtensions.getProjectHandler(projectInfo);
+    // make sure the value we saved in projectinfo is an integer
+    let pingTimeoutInt: number = parseInt(statusPingTimeout);
+
+    if (!pingTimeoutInt) {
```
Comment (rajivnathan) [inline src/pfe/file-watcher/server/src/projects/projectSpecifications.ts]: Think we should use `if(isNaN(pingTimeoutInt))` here instead. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt

File: src/pfe/file-watcher/server/src/projects/projectSpecifications.ts [hunk 3]
```diff
@@ -797,6 +801,61 @@ const changeMavenProperties = async function (mavenProperties: any, operation: O
...
+ *
+ * @returns Promise<any>
+ */
+const changeStatusPingTimeout = async function (statusPingTimeout: string, operation: Operation): Promise<any> {
+
+    const projectInfo: ProjectInfo = operation.projectInfo;
+    const projectID = projectInfo.projectID;
+    const projectHandler = await projectExtensions.getProjectHandler(projectInfo);
+    // make sure the value we saved in projectinfo is an integer
+    let pingTimeoutInt: number = parseInt(statusPingTimeout);
```
Comment (rajivnathan) [inline src/pfe/file-watcher/server/src/projects/projectSpecifications.ts]: When I was reading https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt for the comment below regarding isNaN usage I noticed that if you don't provide radix 10 then it's possible to parse the number incorrectly if the user enters a timeout that starts with `0` or `0x`. We should probably use `parseInt(statusPingTimeout, 10);` to ensure it works the way we expect.

File: src/pfe/file-watcher/server/src/controllers/projectStatusController.ts [hunk 4]
```diff
@@ -470,9 +471,18 @@ function pingInTransitApplications(): void {
...
                                 let newMsg = stateInfo.error;
+
                                 let pingCount = pingCountMap.get(projectID);
-                                // 30 pings = 1 min timeout
-                                const pingCountLimit = 30;
+                                const projectHandler = await projectExtensions.getProjectHandler(projectInfo);
+                                let pingCountLimit;
+                                if (projectInfo.statusPingTimeout) {
+                                    pingCountLimit = projectInfo.statusPingTimeout;
+                                } else if ( projectHandler.getDefaultPingTimeout ) {
```
[the remark this answers is not in the document]
Comment (yangcao77) [reply src/pfe/file-watcher/server/src/controllers/projectStatusController.ts]: updated.

File: src/pfe/file-watcher/server/src/projects/projectSpecifications.ts (hunk 2 above)
Comment (yangcao77) [reply src/pfe/file-watcher/server/src/projects/projectSpecifications.ts]: updated

File: src/pfe/file-watcher/server/src/projects/projectSpecifications.ts (hunk 3 above)
Comment (yangcao77) [reply src/pfe/file-watcher/server/src/projects/projectSpecifications.ts]: updated

File: src/pfe/file-watcher/server/src/controllers/projectStatusController.ts [hunk 5]
```diff
@@ -180,10 +185,11 @@ export async function updateStatus(req: IUpdateStatusParams): Promise<IUpdateSta
  * @param appImageLastBuild <Optional | String> - The last app image timestamp.
  * @param buildImageLastBuild <Optional | String> - The last build image timestamp.
  * @param translatedMsg <Optional | String> - The translated new message.
+ * @param isFirstTimePing <Optional | Boolean> - If it's the first time pinging the app
  *
  * @returns Promise<void>
  */
-export async function updateProjectStatus(type: string, projectID: string, status: string, msg: string, appImageLastBuild?: string, buildImageLastBuild?: string, translatedMsg?: string, detailedAppStatus?: DetailedAppStatus): Promise<void> {
+export async function updateProjectStatus(type: string, projectID: string, status: string, msg: string, appImageLastBuild?: string, buildImageLastBuild?: string, translatedMsg?: string, detailedAppStatus?: DetailedAppStatus, isFirstTimePing?: boolean): Promise<void> {
```
[the remark this answers is not in the document]
Comment (rajivnathan) [reply src/pfe/file-watcher/server/src/controllers/projectStatusController.ts]: As discussed, going with a different solution.

**Pull request merged**